### PR TITLE
Feature/complex property selection

### PR DESCRIPTION
**Pull request by amar-c8y**
in this feature user can select any one of complex property item
exp:- location property has Lat, alt, lang and in this property user can select which one they wanted to see.

### CODE REVIEW

File: asset-properties-widget/component/asset-properties-config/asset-property-item-selector/asset-property-item-selector.component.ts [hunk 1]
```diff
@@ -16,18 +16,64 @@ export class assetPropertyItemSelectorCtrlComponent {
...
       this.selectedProperty.push(cloneDeep(property));
+      this.selectOrUnselectChildren(property, true);
     } else {
-      const removeIndex = this.selectedProperty
+      this.removeUnselectedProperties(property);
+      this.selectOrUnselectChildren(property, false);
+    }
+  }
+
+  selectOrUnselectChildren(selectedProperty, action){
```
Comment (ghost) [inline asset-properties-widget/component/asset-properties-config/asset-property-item-selector/asset-property-item-selector.component.ts]:
```suggestion
  selectOrUnselectChildren(selectedProperty, active){
```

File: asset-properties-widget/component/asset-properties-config/asset-property-item-selector/asset-property-item-selector.component.ts [hunk 2]
```diff
@@ -16,18 +16,64 @@ export class assetPropertyItemSelectorCtrlComponent {
...
+          object['keyPath'] = [property.name];
+          object.keyPath.push(key);
+          constructCustomProperties.push(object);
+        });
+
+      } else {
+        constructProperties.push(property);
+      }
+    });
+    return constructProperties.concat(constructCustomProperties);
```
Comment (ghost) [inline asset-properties-widget/component/asset-properties-config/asset-property-item-selector/asset-property-item-selector.component.ts]: is the requirement to show all simple properties first and then all complex properties?

File: asset-properties-widget/component/asset-properties-config/asset-property-item-selector/asset-property-item-selector.component.ts [hunk 3]
```diff
@@ -16,18 +16,64 @@ export class assetPropertyItemSelectorCtrlComponent {
...
   selectedProperty: IManagedObject[] = [];
   search: string = '';
+  properties: object = [];
+
+  ngOnInit(): void {
+   this.customProperties = cloneDeep(this.constructCustomProperties());
+  }
+
+  constructCustomProperties(): IManagedObject[] {
+    const constructProperties: IManagedObject[] = [];
```
Comment (ghost) [inline asset-properties-widget/component/asset-properties-config/asset-property-item-selector/asset-property-item-selector.component.ts]:
```suggestion
    const simpleProperties: IManagedObject[] = [];
```

File: asset-properties-widget/component/asset-properties-config/asset-property-item-selector/asset-property-item-selector.component.ts [hunk 4]
```diff
@@ -16,18 +16,64 @@ export class assetPropertyItemSelectorCtrlComponent {
...
+    const constructCustomProperties: IManagedObject[] = [];
```
Comment (ghost) [inline asset-properties-widget/component/asset-properties-config/asset-property-item-selector/asset-property-item-selector.component.ts]:
```suggestion
    const complexProperties: IManagedObject[] = [];
```

File: asset-properties-widget/component/asset-properties-view/asset-properties-view.component.ts [hunk 5]
```diff
@@ -51,6 +53,48 @@ export class AssetPropertiesViewComponent implements OnInit {
...
 
+  constructComplexPropertyKeys(){
+    const customizedProperty =[];
+    this.properties.forEach(element => {
+      if(element.keyPath && element.active){
+        const property = this.properties.find((prop) => prop.name === element.keyPath?.[0]) || this.customProperties.find((prop) => prop.name === element.keyPath?.[0]);
+        if(property){
+          if(!property.isHide){
+            property.isHide = true;
+            property.active = true;
```
Comment (ghost) [inline asset-properties-widget/component/asset-properties-view/asset-properties-view.component.ts]: why is both hide and active are set to true? How can something be active and hidden at the same time?

**Review by ghost — CHANGES_REQUESTED** [review]

File: asset-properties-widget/component/asset-properties-config/asset-property-item-selector/asset-property-item-selector.component.ts (hunk 2 above)
Comment (amar-c8y) [reply asset-properties-widget/component/asset-properties-config/asset-property-item-selector/asset-property-item-selector.component.ts]: yes, and same behaviour in older property widget also.

File: asset-properties-widget/component/asset-properties-view/asset-properties-view.component.ts (hunk 5 above)
Comment (amar-c8y) [reply asset-properties-widget/component/asset-properties-view/asset-properties-view.component.ts]: I have changed that flag name isHide to isParentKeySelected, this flag is required when selecting complex property child item key.

**Review by ghost — APPROVED** [review]

**Review by srinath0707 — APPROVED** [review]

**Pull request merged**